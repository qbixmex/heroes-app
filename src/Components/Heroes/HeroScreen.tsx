import { Navigate, useParams } from "react-router-dom";
import getHeroById from "../../Selectors/getHeroById";


const HeroScreen = () => {
  const { id } = useParams();

  const hero = id && getHeroById( id );

  if( !hero ) {
    return <Navigate to="/" />
  }

  return (
    <>
      <h1 className="text-center display-1 green">
        { hero.superhero }
      </h1>
    </>
  );
};

export default HeroScreen;
