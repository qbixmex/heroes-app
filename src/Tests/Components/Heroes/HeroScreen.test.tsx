import { mount } from "enzyme";
import { Route, Routes, MemoryRouter } from 'react-router-dom';
import HeroScreen from "../../../Components/Heroes/HeroScreen";

const mockNavigate = jest.fn()

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate
}));

//@ts-ignore
describe("Test on HeroScreen", () => {
  
  test("Should not show HeroScreen if url is empty", () => {
    const message = "There's no hero screen";
  
    const wrapper = mount(
      <MemoryRouter initialEntries={ ["/hero"] }>
        <Routes>
          <Route path="/hero" element={ <HeroScreen /> } />
          <Route path="/" element={ <h1>{ message }</h1> } />
        </Routes>
      </MemoryRouter>
    );

    expect( wrapper.find("h1").text().trim() ).toBe( message );
  });

  test("Should a hero if param exists and it was founded", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={ ["/hero/marvel-spider"] }>
        <Routes>
          <Route path="/hero/:id" element={ <HeroScreen /> } />
        </Routes>
      </MemoryRouter>
    );

    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find("h1").text().trim() ).toBe("Spider Man");
  });

  test("Should return back to previous screen", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={ ["/hero/marvel-spider"] }>
        <Routes>
          <Route path="/hero/:id" element={ <HeroScreen /> } />
        </Routes>
      </MemoryRouter>
    );

    const mockEvent = {} as React.MouseEvent<HTMLButtonElement>;
    const handleClick = wrapper.find("button").prop<React.MouseEventHandler>("onClick");

    handleClick(mockEvent);

    expect( mockNavigate ).toHaveBeenCalledWith( -1 );
  });

  test("Should show no Hero Page if doesn't have a hero", () => {
    const message = "No Hero Page";
    const wrapper = mount(
      <MemoryRouter initialEntries={ ["/hero/marvel-spider123"] }>
        <Routes>
          <Route path="/hero/:id" element={ <HeroScreen /> } />
          <Route path="/" element={ <h1>{ message }</h1> } />
        </Routes>
      </MemoryRouter>
    );

    expect( wrapper.text() ).toBe( message );
  });
});
