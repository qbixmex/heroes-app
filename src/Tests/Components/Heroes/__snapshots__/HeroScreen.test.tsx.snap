// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on HeroScreen Should a hero if param exists and it was founded 1`] = `
<div
  className="row mt-5"
>
  <div
    className="col-12 col-md-4"
  >
    <img
      alt="Spider Man"
      className="img-thumbnail animate__animated animate__fadeInLeft"
      src="/assets/marvel-spider.jpg"
      title="Spider Man"
    />
  </div>
  <div
    className="col-12 col-md-8 animate__animated animate__fadeInRight"
  >
    <h1
      className="text-center display-4 green mt-4 mt-md-0 mb-4"
    >
      Spider Man
    </h1>
    <table
      className="table table-dark"
    >
      <tbody>
        <tr>
          <th>
            Alter Ego
          </th>
          <td>
            Peter Parker
          </td>
        </tr>
        <tr>
          <th>
            Publisher
          </th>
          <td>
            Marvel Comics
          </td>
        </tr>
        <tr>
          <th>
            First Appearance
          </th>
          <td>
            Amazing Fantasy #15
          </td>
        </tr>
        <tr />
      </tbody>
    </table>
    <div
      className="text-end"
    >
      <button
        className="btn btn-outline-primary"
        onClick={[Function]}
      >
        Back
      </button>
    </div>
  </div>
</div>
`;
